File: src/leetcode/hash_maps_and_sets/2215_find_difference_of_two_arrays.ts
// LeetCode 75
// https://leetcode.com/problems/find-the-difference-of-two-arrays/description/?envType=study-plan-v2&envId=leetcode-75

// Time complexity: O(n + m)
// Space complexity: O(n + m)

const findDifference = (nums1: number[], nums2: number[]): number[][] => {
  const nums1Set = new Set(nums1);
  const nums2Set = new Set(nums2);

  const nums1Nums2Diff = [...nums1Set].filter((num) => !nums2Set.has(num));
  const nums2Nums1Diff = [...nums2Set].filter((num) => !nums1Set.has(num));

  return [nums1Nums2Diff, nums2Nums1Diff];
};

export default findDifference;

File: src/leetcode/hash_maps_and_sets/2215_find_difference_of_two_arrays.rb
# frozen_string_literal: true

# LeetCode 75
# https://leetcode.com/problems/find-the-difference-of-two-arrays/description/?envType=study-plan-v2&envId=leetcode-75

# Time complexity: O(n + m)
# Space complexity: O(n + m)

# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Integer[][]}
def find_difference_set(nums1, nums2)
  nums1_set = nums1.to_set
  nums2_set = nums2.to_set

  [(nums1_set - nums2_set).to_a, (nums2_set - nums1_set).to_a]
end

def find_difference_uniq(nums1, nums2)
  nums1_uniq = nums1.uniq
  nums2_uniq = nums2.uniq

  [nums1_uniq - nums2_uniq, nums2_uniq - nums1_uniq]
end

def find_difference_bitwise_or(nums1, nums2)
  uniq_combined = nums1 | nums2

  [uniq_combined - nums2, uniq_combined - nums1]
end
